pkg/templater/expander.go

```diff
@@ -0,0 +1,157 @@
+package templater
+
+import (
+	"github.com/treeverse/lakefs/pkg/auth"
+	auth_model "github.com/treeverse/lakefs/pkg/auth/model"
+	config "github.com/treeverse/lakefs/pkg/config"
+
+	"context"
+	"errors"
+	"fmt"
+	"path"
+	//	html_template "html/template"
+	"io"
+	"io/fs"
+	"strings"
+	"text/template"
+)
+
+const (
+	// Prefix inside templates FS to access actual contents.
+	embeddedContentPrefix = "content"
+)
+
+var (
+	ErrTemplateNotFound     = errors.New("template not found")
+	ErrPathTraversalBlocked = errors.New("path traversal blocked")
+)
+
+type AuthService interface {
+	auth.Authorizer
+	auth.CredentialsCreator
+}
+
+type ControlledParams struct {
+	Ctx  context.Context
+	Auth AuthService
+	// User is the user expanding the template.
+	User *auth_model.User
+}
+
+type UncontrolledData struct {
+	// UserName is the name of the executing user.
+	Username string
+	// Query is the (parsed) querystring of the HTTP access.
+	Query map[string]string
+}
+
+// Params parametrizes a single template expansion.
+type Params struct {
+	// Controlled is the data visible to functions to control expansion.
+	// It is _not_ directly visible to templates for expansion.
+	Controlled *ControlledParams
+	// Data is directly visible to templates for expansion, with no
+	// authorization required.
+	Data *UncontrolledData
+}
+
+// Expander is a template that may be expanded as requested by users.
+type Expander interface {
+	// Prepare checks whether the template expansion will succeed.  Call
+	// it before Expand to ensure nothing is written when expansion fails.
+	Prepare(params *Params) error
```

Comment:
Why do you need this two step operation?
Can't you use `Expand` to try and `Execute` to a temporary `bytes.Buffer{}`, return error if it doesn't succeed, and write to the given Writer if it does succeed?

Reply:
We could, but I am not sure that it would be better than this.  One disadvantage: I worry that the buffer might be large.  As an example, suppose someone abused template to serve a large object (maybe even static, say an image).  Then we would hold it in process memory, which is not fun for the process.

Are there advantages other than simpler implementation?

Reply:
I do think that simpler implementation (thus better maintainability) is important enough, but the concerns you're raising are more important.
Regardless you can still collapse it to a single step process and do the validation step as a first step of the expansion... I think that it's not the role of the client to check if the template is expandable or not.

Reply:
Sounds good!  I shall rework the logic from `service.Expand` into a single call here.